File: shoe-shop-frontend-app/src/app/services/review.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class ReviewService {
  readonly baseUrl = 'http://localhost:8080/api/v1/review';
  constructor(private http: HttpClient) { }

  findAllReviewByProductDetail(idProductDetail: number): Observable<any> {
    return this.http.get(`${this.baseUrl}/${idProductDetail}`);
  }

  countReviewByProductDetail(idProductDetail: number): Observable<any> {
    return this.http.get(`${this.baseUrl}/count/${idProductDetail}`);
  }

  countReviewByRatingAndProductDetail(rating: number, idProductDetail: number): Observable<any> {
    return this.http.get(`${this.baseUrl}/count-per-rating/${rating}/${idProductDetail}`);
  }

  checkProductReviewed(orderId: number, productDetailId: number, userId: number): Observable<any> {
    return this.http.get(`${this.baseUrl}/check-reviewed/${orderId}/${productDetailId}/${userId}`);
  }

  saveReview(review: any, file: File): Observable<any> {
    const formData: FormData = new FormData();
    formData.append('review', new Blob([JSON.stringify(review)], {type: 'application/json'}));
    formData.append('file', file, file.name);

    return this.http.post(this.baseUrl, formData);
  }
}

File: shoe-shop-backend-app/src/main/java/root/app/service/ReviewService.java
package root.app.service;

import org.springframework.web.multipart.MultipartFile;
import root.app.entity.Review;
import java.util.List;

public interface ReviewService {

    List<Review> findReviewByProductDetail(Long productDetailId);

    Integer countReviewByProductDetail(Long productDetailId);

    Integer countReviewByRatingAndProductDetail(Integer rating, Long productDetailId);

    Boolean checkProductReviewed(Long orderId, Long productDetailId, Long userId);

    Double getAverageRating(Long idProductDetail);

    Review saveReview(Review review, MultipartFile file);

    void deleteReview(Long idReview);
}
